api/src/services/setup/upgrade-log.js

```diff
@@ -0,0 +1,231 @@
+const db = require('../../db');
+const logger = require('../../logger');
+
+/**
+ * @typedef {Object} UpgradeLog
+ * @property {string} _id
+ * @property {string} user - username of user initiating the upgrade (eg. "mary")
+ * @property {string} from_version - eg "4.0.3"
+ * @property {string} to_version - eg "4.1.2"
+ * @property {number} start_date - timestamp when upgrade was initiated
+ * @property {number} updated_date - timestamp when the log was last updated
+ * @property {string} state - latest upgrade state
+ * @property {Array<Object>} state_history
+ * @property {number} state_history[].date
+ * @property {string} state_history[].state
+ * @property {any} state_history[].details
+ */
+
+const UPGRADE_LOG_STATES = {
+  INITIATED: 'initiated',
+  STAGED: 'staged',
+  INDEXING: 'indexing',
+  INDEXED: 'indexed',
+  COMPLETING: 'completing',
+  FINALIZING: 'finalizing',
+  FINALIZED: 'finalized',
+  COMPLETE: 'complete',
+  ABORTED: 'aborted',
+  ERRORED: 'errored',
+};
+
+let currentUpgradeLogId;
+
+const isFinalState = (state) => {
+  return state === UPGRADE_LOG_STATES.FINALIZED ||
+         state === UPGRADE_LOG_STATES.ABORTED ||
+         state === UPGRADE_LOG_STATES.ERRORED;
+};
+
+const UPGRADE_LOG_NAME = 'upgrade_log';
+
+const getUpgradeLogId = (version, startDate) => `${UPGRADE_LOG_NAME}:${startDate}:${version}`;
+
+const getLatestUpgradeLog = async () => {
+  const now = new Date().getTime();
+  const results = await db.medicLogs.allDocs({
+    startkey: `${UPGRADE_LOG_NAME}:${now}:`,
```

Comment:
Adding `${now}` doesn't change anything as far as I can tell, because we shouldn't have any logs from the future. Can we just have...

```suggestion
    startkey: `${UPGRADE_LOG_NAME}:`,
```

Reply:
It's just so somebody manually creating a upgrade_log from 2032 will screw up all of our upgrade attempts :D 
I'd rather keep it.